### PR TITLE
create test for mysql.go (part1)

### PR DESCRIPTION
**Pull request by vankichi**
Signed-off-by: vankichi <kyukawa315@gmail.com>

### Description:

I created the test for a part of the method at `internal/db/rdb/mysql.go`.
Because, It will be create large diff when I created test for all method in `mysql.go`.
And also, I created the dbr mock code for test instead of connecting real MySQL database.

And more, I refactored code for test.
  - Use `pctx` instead of `ctx` which is parent of `pctx` at `Ping()`.
     It is just for testing and will not affect current logic.

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/db/rdb/mysql/mysql_test.go [hunk 1]
```diff
@@ -636,82 +677,275 @@ func Test_mySQLClient_GetMeta(t *testing.T) {
...
+				afterFunc: func(args) {
+					go func() {
+						cancel()
+					}()
+				},
+			}
+		}(),
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
+			uuid := "vdaas-01"
```
[the remark this answers is not in the document]
Comment (vankichi) [reply internal/db/rdb/mysql/mysql_test.go:745]: How should I do ? 😅

File: internal/db/rdb/mysql/mysql_test.go [hunk 2]
```diff
@@ -19,18 +19,35 @@ package mysql
...
+var (
+	mysqlComparatorOptions = []comparator.Option{
+		comparator.AllowUnexported(mySQLClient{}),
+		comparator.Comparer(func(x, y dbr.DBR) bool {
+			return reflect.DeepEqual(x, y)
+		}),
+	}
+)
+
+func TestMain(m *testing.M) {
```
[the remark this answers is not in the document]
Comment (vankichi) [reply internal/db/rdb/mysql/mysql_test.go:36]: How should I do ? 😅

File: internal/db/rdb/mysql/mysql_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/db/rdb/mysql/mysql_test.go:745]: I think we should ignore it

File: internal/db/rdb/mysql/mysql_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/db/rdb/mysql/mysql_test.go:36]: ignore it! :)

File: internal/db/rdb/mysql/mysql_test.go [hunk 3]
```diff
@@ -57,36 +64,65 @@ func TestNew(t *testing.T) {
...
+		func() test {
+			n := "vald"
+			opts := []Option{
+				WithName(n),
+			}
+			m := new(mySQLClient)
+			for _, opt := range append(defaultOpts, opts...) {
+				_ = opt(m)
+			}
+			m.dbr = dbr.New()
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:89]:
```suggestion
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 4]
```diff
@@ -151,80 +186,172 @@ func Test_mySQLClient_Open(t *testing.T) {
...
+					session: &dbr.MockSession{
+						PingContextFunc: func(ctx context.Context) error {
+							return nil
+						},
+					},
+					connected: func() atomic.Value {
+						v := new(atomic.Value)
+						v.Store(false)
+						return *v
+					}(),
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:220]:
```suggestion
					connected: func() (v atomic.Value) {
						v.Store(false)
						return
					}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 5]
```diff
@@ -151,80 +186,172 @@ func Test_mySQLClient_Open(t *testing.T) {
...
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:281]:
```suggestion
					connected: func() (v atomic.Value) {
						v.Store(false)
						return
					}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 6]
```diff
@@ -151,80 +186,172 @@ func Test_mySQLClient_Open(t *testing.T) {
...
+					initialPingTimeLimit: 1000 * time.Microsecond,
+					initialPingDuration:  10 * time.Microsecond,
+					connMaxLifeTime:      1 * time.Microsecond,
+					maxOpenConns:         10,
+					maxIdleConns:         10,
+					connected: func() atomic.Value {
+						v := new(atomic.Value)
+						v.Store(false)
+						return *v
+					}(),
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:333]:
```suggestion
					connected: func() (v atomic.Value) {
						v.Store(false)
						return
					}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 7]
```diff
@@ -311,80 +420,88 @@ func Test_mySQLClient_Ping(t *testing.T) {
...
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
+			err := errors.New("error")
+			return test{
+				name: "returns error when session.PingContext returns error",
+				args: args{
+					ctx: ctx,
+				},
+				fields: fields{
+					initialPingTimeLimit: 100 * time.Microsecond,
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:456]:
```suggestion
					initialPingTimeLimit: 30 * time.Microsecond,
```

I think if you set to `100`, your screen fill error output log.
so could you please set to `30` or `15`?

File: internal/db/rdb/mysql/mysql_test.go [hunk 8]
```diff
@@ -460,91 +541,59 @@ func Test_mySQLClient_Close(t *testing.T) {
...
+			args: args{
+				ctx: context.Background(),
+			},
+			fields: fields{
+				session: &dbr.MockSession{},
+				connected: func() atomic.Value {
+					v := new(atomic.Value)
+					v.Store(false)
+					return *v
+				}(),
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:569]:
```suggestion
				connected: func() (v atomic.Value) {
					v.Store(false)
					return
				}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 9]
```diff
@@ -460,91 +541,59 @@ func Test_mySQLClient_Close(t *testing.T) {
...
+				session: &dbr.MockSession{
+					CloseFunc: func() error {
+						return nil
+					},
+				},
+				connected: func() atomic.Value {
+					v := new(atomic.Value)
+					v.Store(true)
+					return *v
+				}(),
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:588]:
```suggestion
				connected: func() (v atomic.Value) {
					v.Store(true)
					return
				}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 10]
```diff
@@ -636,82 +667,273 @@ func Test_mySQLClient_GetMeta(t *testing.T) {
...
+				args: args{
+					ctx:  ctx,
+					uuid: "",
+				},
+				fields: fields{
+					connected: func() atomic.Value {
+						v := new(atomic.Value)
+						v.Store(false)
+						return *v
+					}(),
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:683]:
```suggestion
					connected: func() (v atomic.Value) {
						v.Store(false)
						return
					}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 11]
```diff
@@ -636,82 +667,273 @@ func Test_mySQLClient_GetMeta(t *testing.T) {
...
+								return 0, err
+							}
+							return m
+						},
+					},
+					connected: func() atomic.Value {
+						v := new(atomic.Value)
+						v.Store(true)
+						return *v
+					}(),
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:726]:
```suggestion
					connected: func() (v atomic.Value) {
						v.Store(true)
						return
					}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 12]
```diff
@@ -636,82 +667,273 @@ func Test_mySQLClient_GetMeta(t *testing.T) {
...
+								return 0, errors.New("not found")
+							}
+							return s
+						},
+					},
+					connected: func() atomic.Value {
+						v := new(atomic.Value)
+						v.Store(true)
+						return *v
+					}(),
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:779]:
```suggestion
					connected: func() (v atomic.Value) {
						v.Store(true)
						return
					}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 13]
```diff
@@ -636,82 +667,273 @@ func Test_mySQLClient_GetMeta(t *testing.T) {
...
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:842]:
```suggestion
					connected: func() (v atomic.Value) {
						v.Store(true)
						return
					}(),
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 14]
```diff
@@ -636,82 +667,273 @@ func Test_mySQLClient_GetMeta(t *testing.T) {
...
+								return 0, errors.New("error")
+							}
+							return s
+						},
+					},
+					connected: func() atomic.Value {
+						v := new(atomic.Value)
+						v.Store(true)
+						return *v
+					}(),
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:912]:
```suggestion
					connected: func() (v atomic.Value) {
						v.Store(true)
						return
					}(),
```

**Review by hlts2 — CHANGES_REQUESTED** [review]

File: internal/db/rdb/mysql/dbr/dbr_mock.go [hunk 15]
```diff
@@ -0,0 +1,126 @@
+package dbr
```
Comment (kevindiu) [inline internal/db/rdb/mysql/dbr/dbr_mock.go:1]: should we name this file to `dbr_mock_test.go` to ignore this file to generate the test file and ignore the test coverage?

File: internal/db/rdb/mysql/mysql_test.go [hunk 16]
```diff
@@ -151,80 +186,172 @@ func Test_mySQLClient_Open(t *testing.T) {
...
+					go func() {
+						cancel()
+					}()
+				},
+			}
+		}(),
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
+			return test{
+				name: "returns error when dbr.Open failed",
```
Comment (kevindiu) [inline internal/db/rdb/mysql/mysql_test.go:311]: can you also add a test case of ping error?

File: internal/db/rdb/mysql/mysql.go [hunk 17]
```diff
@@ -143,13 +150,13 @@ func (m *mySQLClient) Ping(ctx context.Context) (err error) {
...
 			}
-			cerr := ctx.Err()
+			cerr := pctx.Err()
 			if cerr != nil {
 				err = errors.Wrap(err, cerr.Error())
 			}
 			return err
 		case <-tick.C:
-			err = m.session.PingContext(ctx)
+			err = m.session.PingContext(pctx)
```
[the remark this answers is not in the document]
Comment (vankichi) [reply internal/db/rdb/mysql/mysql.go:159]: this implementation already discussed with author.
We do not have to author review

File: internal/db/rdb/mysql/mysql_test.go (hunk 16 above)
Comment (vankichi) [reply internal/db/rdb/mysql/mysql_test.go:302]: I'm not sure why add this pattern.
ping error test is created in Ping func test.
could you please show me your opinion?

File: internal/db/rdb/mysql/dbr/dbr_mock.go (hunk 15 above)
Comment (vankichi) [reply internal/db/rdb/mysql/dbr/dbr_mock.go:1]: If we should replace mock file to `internal/test/mock`, I'll adopt next PR(mysql test part2).
Is it OK?

Comment (kevindiu) [reply internal/db/rdb/mysql/dbr/dbr_mock.go:16]: ok.
If we do so, do we need to update coding guideline?

File: internal/db/rdb/mysql/mysql_test.go (hunk 16 above)
Comment (kevindiu) [reply internal/db/rdb/mysql/mysql_test.go:298]: My idea was to think about how `dbr.Open()` works. (like black box testing and white box testing)
In our implementation, `dbr.Open()` will ping the db and return the error, but in the test code there are no such test case.

If you think it is ok to not to implement such test case, I think it is ok too

File: internal/db/rdb/mysql/mysql_test.go [hunk 18]
```diff
@@ -636,82 +643,259 @@ func Test_mySQLClient_GetMeta(t *testing.T) {
...
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
+			uuid := "vdaas-01"
+			m := &meta{
+				ID:     1,
+				UUID:   uuid,
+				Vector: []byte("0.1,0.2"),
+			}
+			p := []podIP{
+				podIP{
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/db/rdb/mysql/mysql_test.go:832]: please fix it 🤔

File: internal/db/rdb/mysql/dbr/dbr_mock.go (hunk 15 above)
Comment (vankichi) [reply internal/db/rdb/mysql/dbr/dbr_mock.go:16]: I think if we use mock for test, we should write handling test with test at test section in coding guideline

**Review by hlts2 — APPROVED** [review]
LGTM

File: internal/db/rdb/mysql/dbr/dbr_mock.go (hunk 15 above)
Comment (kevindiu) [reply internal/db/rdb/mysql/dbr/dbr_mock.go:16]: lets discuss it in different place :)

**Review by kevindiu — APPROVED** [review]
LGTM

**Pull request merged**